lib/modules/datasource/maven/util.ts

```diff
@@ -130,6 +180,64 @@ export async function checkHttpResource(
   }
 }
 
+async function checkS3Resource(
+  pkgUrl: URL | string
+): Promise<HttpResourceCheckResult> {
+  try {
+    const s3Url = parseS3Url(pkgUrl.toString());
+    // istanbul ignore next
+    if (s3Url === null) {
+      return 'error';
+    }
+    const response = await getS3Client().headObject(s3Url);
+    // istanbul ignore next
+    if (response.DeleteMarker) {
+      return 'not-found';
+    }
+    if (response.LastModified) {
+      return response.LastModified;
+    }
+    return 'found';
+  } catch (err) {
+    if (isS3NotFound(err)) {
+      return 'not-found';
+    } /* istanbul ignore else */ else {
+      // istanbul ignore next
+      logger.debug(
+        { pkgUrl, name: err.name, message: err.message },
+        `Can't check S3 resource existence`
+      );
+    }
+    // istanbul ignore next
+    return 'error';
+  }
+}
+
+export async function checkResource(
+  http: Http,
+  pkgUrl: URL | string
+): Promise<HttpResourceCheckResult> {
+  const parsedUrl = typeof pkgUrl === 'string' ? parseUrl(pkgUrl) : pkgUrl;
+  // istanbul ignore next
+  if (parsedUrl === null) {
+    return 'error';
+  }
+  switch (parsedUrl.protocol) {
+    case 'http:':
+    case 'https:':
+      return await checkHttpResource(http, parsedUrl);
+    case 's3:':
+      return await checkS3Resource(parsedUrl);
+    /* istanbul ignore next */
```

Comment:
This switch statement is the same as in `downloadMavenXml`. If that falls through, then this switch will logically never fall through. I've added an explanation of that.

Reply:
even when other code is doing the same it's not our best practices.

we now like to avoid any coverage skips, as we had too many issues in skipped code blocks.

this should be very easy to test. you can call this function directly in tests and pass a ssh url to trigger that case.

so only ignore code coverage if it's not testable or it's really too hard to create a synthetic test for something which shouldn't happen at runtime.

@rarkins @HonkingGoose can we add something about this to our devel docs, so we can point out? or better an auto label comment for easier usage.

Reply:
- #15138